sdks/java/harness/src/main/java/org/apache/beam/fn/harness/debug/OutputSampler.java

```diff
@@ -91,10 +94,13 @@ public void sample(T element) {
   /**
    * Clears samples at end of call. This is to help mitigate memory use.
    *
+   * <p>This method is invoked by a thread handling a data sampling request in parallel to any calls
+   * to {@link #sample}.
+   *
    * @return samples taken since last call.
    */
-  public List<byte[]> samples() {
-    List<byte[]> ret = new ArrayList<>();
+  public List<BeamFnApi.SampledElement> samples() throws IOException {
+    List<BeamFnApi.SampledElement> ret = new ArrayList<>();
 
     // Serializing can take a lot of CPU time for larger or complex elements. Copy the array here
     // so as to not slow down the main processing hot path.
```

Comment:
Since your only ever accessing buffer under a synchronized block it will be better to do a buffer swap then to copy the buffer.

e.g.
```
    List<T> bufferToSend;
    synchronized (this) {
      bufferToSend = buffer;
      buffer = new ArrayList(maxElements);
      resampleIndex = 0;
    }
```

Saves on copying the elements and clearing the original.

Reply:
Done!